### PR TITLE
enable mkl_gemm_f16f16f32 in cpublas::gemm

### PR DESCRIPTION
**Pull request by CaoE**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118368
* __->__ #118367

### CODE REVIEW

**Review by jgong5 — APPROVED** [review]

**Comment by CaoE** [comment]
@nikitaved Could you please help review this PR ? Thanks.

**Comment by CaoE** [comment]
@cpuhrsch Could you please help review this PR ? Thanks.

File: aten/src/ATen/native/CPUBlas.h [hunk 1]
```diff
@@ -88,6 +88,15 @@ void gemm(
     float beta,
     at::Half *c, int64_t ldc);
 
+void gemm(
```
Comment (cpuhrsch) [inline aten/src/ATen/native/CPUBlas.h:91]: I understand you mainly want this for flashattention, but this seems like a general purpose matmul CPU kernel for float16? Would it make sense to also expose this to torch.mm for CPU and extend test coverage there too? That way you have more test coverage for flash attention too.

Comment (CaoE) [reply aten/src/ATen/native/CPUBlas.h:91]: Thanks for your comments. mkl_gemm_f16f16f32 is only added for flash attention. It is same as https://github.com/pytorch/pytorch/pull/107196. In the attention kernel, q @ k.T = attn, q and k will be given as float16 and attn is float32.

**Review by cpuhrsch — APPROVED** [review]

**Comment by cpuhrsch** [comment]
cc @drisspg

**Pull request closed without merging**